File: integrations/api-sdks/javascript/codeguardian-sdk.js
/**
 * CodeGuardian JavaScript SDK
 * Official JavaScript/TypeScript client library for CodeGuardian API
 */

const axios = require('axios');
const fs = require('fs').promises;
const path = require('path');
const glob = require('glob');

/**
 * Severity levels for code issues
 */
const Severity = {
    HIGH: 'high',
    MEDIUM: 'medium',
    LOW: 'low'
};

/**
 * CodeGuardian API Error Classes
 */
class CodeGuardianError extends Error {
    constructor(message) {
        super(message);
        this.name = 'CodeGuardianError';
    }
}

class AuthenticationError extends CodeGuardianError {
    constructor(message = 'Invalid API key') {
        super(message);
        this.name = 'AuthenticationError';
    }
}

class RateLimitError extends CodeGuardianError {
    constructor(message = 'Rate limit exceeded') {
        super(message);
        this.name = 'RateLimitError';
    }
}

class AnalysisError extends CodeGuardianError {
    constructor(message = 'Analysis failed') {
        super(message);
        this.name = 'AnalysisError';
    }
}

/**
 * Data classes for API responses
 */
class CodeIssue {
    constructor({
        lineNumber,
        column,
        severity,
        category,
        message,
        suggestion = null,
        fix = null
    }) {
        this.lineNumber = lineNumber;
        this.column = column;
        this.severity = severity;
        this.category = category;
        this.message = message;
        this.suggestion = suggestion;
        this.fix = fix;
    }
}

class AnalysisResult {
    constructor({
        overallScore,
        securityScore,
        performanceScore,
        maintainabilityScore,
        issues,
        suggestions,
        analysisTime
    }) {
        this.overallScore = overallScore;
        this.securityScore = securityScore;
        this.performanceScore = performanceScore;
        this.maintainabilityScore = maintainabilityScore;
        this.issues = issues;
        this.suggestions = suggestions;
        this.analysisTime = analysisTime;
    }
}

class ExplanationResult {
    constructor({
        description,
        suggestions,
        reasoning,
        examples = []
    }) {
        this.description = description;
        this.suggestions = suggestions;
        this.reasoning = reasoning;
        this.examples = examples;
    }
}

/**
 * CodeGuardian API Client
 * 
 * Provides access to CodeGuardian's AI-powered code analysis and mentorship features.
 * 
 * @example
 * const client = new CodeGuardianClient({ apiKey: 'your-api-key' });
 * const result = await client.analyzeCode("console.log('hello world')", { language: 'javascript' });
 * console.log(`Score: ${result.overallScore}/100`);
 */
class CodeGuardianClient {
    /**
     * Initialize CodeGuardian client
     * 
     * @param {Object} options - Configuration options
     * @param {string} options.apiKey - Your CodeGuardian API key
     * @param {string} [options.apiUrl='https://codeguardian-api.onrender.com'] - API base URL
     * @param {number} [options.timeout=30000] - Request timeout in milliseconds
     * @param {number} [options.maxRetries=3] - Maximum number of retry attempts
     */
    constructor({
        apiKey,
        apiUrl = 'https://codeguardian-api.onrender.com',
        timeout = 30000,
        maxRetries = 3
    }) {
        this.apiKey = apiKey;
        this.apiUrl = apiUrl.replace(/\/$/, '');
        this.timeout = timeout;
        this.maxRetries = maxRetries;

        this.client = axios.create({
            baseURL: this.apiUrl,
            timeout: this.timeout,
            headers: {
                'Authorization': `Bearer ${apiKey}`,
                'Content-Type': 'application/json',
                'User-Agent': 'CodeGuardian-JavaScript-SDK/1.0.0'
            }
        });

        this._setupInterceptors();
    }

    /**
     * Set up axios interceptors for error handling and retries
     * @private
     */
    _setupInterceptors() {
        this.client.interceptors.response.use(
            response => response,
            async error => {
                const { response, config } = error;

                if (response?.status === 401) {
                    throw new AuthenticationError();
                } else if (response?.status === 429) {
                    // Retry with exponential backoff for rate limits
                    if (!config._retryCount) config._retryCount = 0;
                    
                    if (config._retryCount < this.maxRetries) {
                        config._retryCount++;
                        const delay = Math.pow(2, config._retryCount) * 1000;
                        await this._sleep(delay);
                        return this.client(config);
                    }
                    
                    throw new RateLimitError();
                } else if (response?.status >= 400) {
                    throw new CodeGuardianError(`API error: ${response.status}`);
                }

                // Retry for network errors
                if (!config._retryCount) config._retryCount = 0;
                
                if (config._retryCount < this.maxRetries) {
                    config._retryCount++;
                    await this._sleep(1000);
                    return this.client(config);
                }

                throw new CodeGuardianError(`Request failed: ${error.message}`);
            }
        );
    }

    /**
     * Sleep for specified milliseconds
     * @private
     */
    _sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    /**
     * Analyze code for issues, security vulnerabilities, and improvements
     * 
     * @param {string} code - Source code to analyze
     * @param {Object} [options={}] - Analysis options
     * @param {string} [options.language='javascript'] - Programming language
     * @param {string} [options.filename] - Optional filename for context
     * @param {boolean} [options.mentorship=true] - Enable AI mentorship explanations
     * @param {boolean} [options.securityFocus=true] - Focus on security vulnerabilities
     * @returns {Promise<AnalysisResult>} Analysis results
     * 
     * @throws {AnalysisError} If analysis fails
     * @throws {AuthenticationError} If API key is invalid
     * @throws {RateLimitError} If rate limit is exceeded
     */
    async analyzeCode(code, options = {}) {
        const {
            language = 'javascript',
            filename = null,
            mentorship = true,
            securityFocus = true,
            ...additionalOptions
        } = options;

        const payload = {
            code,
            language,
            fileName: filename,
            options: {
                mentorship,
                securityFocus,
                ...additionalOptions
            }
        };

        const startTime = Date.now();
        
        try {
            const response = await this.client.post('/api/reviews/analyze', payload);
            const analysisTime = Date.now() - startTime;
            
            return this._parseAnalysisResult(response.data, analysisTime);
        } catch (error) {
            if (error instanceof CodeGuardianError) {
                throw error;
            }
            throw new AnalysisError(`Analysis failed: ${error.message}`);
        }
    }

    /**
     * Analyze a code file
     * 
     * @param {string} filePath - Path to the code file
     * @param {Object} [options={}] - Analysis options
     * @param {string} [options.language] - Programming language (auto-detected if not provided)
     * @returns {Promise<AnalysisResult>} Analysis results
     */
    async analyzeFile(filePath, options = {}) {
        try {
            const code = await fs.readFile(filePath, 'utf8');
            const language = options.language || this._detectLanguage(filePath);
            
            return this.analyzeCode(code, {
                ...options,
                language,
                filename: filePath
            });
        } catch (error) {
            if (error.code === 'ENOENT') {
                throw new Error(`File not found: ${filePath}`);
            }
            throw error;
        }
    }

    /**
     * Analyze all code files in a directory
     * 
     * @param {string} directoryPath - Path to the directory
     * @param {Object} [options={}] - Analysis options
     * @param {boolean} [options.recursive=true] - Analyze subdirectories recursively
     * @param {string[]} [options.filePatterns] - File patterns to include
     * @returns {Promise<Object>} Dictionary mapping file paths to analysis results
     */
    async analyzeDirectory(directoryPath, options = {}) {
        const {
            recursive = true,
            filePatterns = [
                '**/*.js', '**/*.ts', '**/*.jsx', '**/*.tsx',
                '**/*.py', '**/*.java', '**/*.cpp', '**/*.c',
                '**/*.cs', '**/*.go', '**/*.rs', '**/*.php', '**/*.rb'
            ],
            ...analysisOptions
        } = options;

        const globOptions = {
            cwd: directoryPath,
            absolute: true,
            ignore: ['**/node_modules/**', '**/dist/**', '**/build/**']
        };

        if (!recursive) {
            globOptions.depth = 1;
        }

        const results = {};
        
        for (const pattern of filePatterns) {
            const files = glob.sync(pattern, globOptions);
            
            for (const filePath of files) {
                try {
                    const result = await this.analyzeFile(filePath, analysisOptions);
                    results[filePath] = result;
                } catch (error) {
                    console.error(`Error analyzing ${filePath}:`, error.message);
                }
            }
        }

        return results;
    }

    /**
     * Get AI explanation and mentorship for code
     * 
     * @param {string} code - Source code to explain
     * @param {Object} [options={}] - Explanation options
     * @param {string} [options.language='javascript'] - Programming language
     * @param {string} [options.mentorshipLevel='detailed'] - Level of detail
     * @returns {Promise<ExplanationResult>} Explanation results
     */
    async explainCode(code, options = {}) {
        const {
            language = 'javascript',
            mentorshipLevel = 'detailed'
        } = options;

        const payload = {
            code,
            language,
            mentorshipLevel
        };

        try {
            const response = await this.client.post('/api/reviews/explain', payload);
            
            return new ExplanationResult({
                description: response.data.description || '',
                suggestions: response.data.suggestions || '',
                reasoning: response.data.reasoning || '',
                examples: response.data.examples || []
            });
        } catch (error) {
            throw new CodeGuardianError(`Explanation failed: ${error.message}`);
        }
    }

    /**
     * Get improvement suggestions for code
     * 
     * @param {string} code - Source code
     * @param {Object} [options={}] - Options
     * @param {string} [options.language='javascript'] - Programming language
     * @param {string} [options.context] - Additional context for suggestions
     * @returns {Promise<string[]>} List of improvement suggestions
     */
    async getSuggestions(code, options = {}) {
        const {
            language = 'javascript',
            context = null
        } = options;

        const payload = {
            code,
            language,
            context
        };

        try {
            const response = await this.client.post('/api/reviews/suggestions', payload);
            return response.data.suggestions || [];
        } catch (error) {
            throw new CodeGuardianError(`Failed to get suggestions: ${error.message}`);
        }
    }

    /**
     * Check code for security vulnerabilities
     * 
     * @param {string} code - Source code to check
     * @param {Object} [options={}] - Options
     * @param {string} [options.language='javascript'] - Programming language
     * @returns {Promise<CodeIssue[]>} List of security issues found
     */
    async checkSecurity(code, options = {}) {
        const { language = 'javascript' } = options;
        
        const result = await this.analyzeCode(code, {
            language,
            securityFocus: true,
            mentorship: false
        });

        // Filter for security-related issues
        return result.issues.filter(issue => 
            issue.category.toLowerCase().includes('security')
        );
    }

    /**
     * Analyze multiple code samples concurrently
     * 
     * @param {Array} codeSamples - Array of {code, language, options} objects
     * @returns {Promise<AnalysisResult[]>} Array of analysis results
     */
    async analyzeMultiple(codeSamples) {
        const promises = codeSamples.map(sample => 
            this.analyzeCode(sample.code, {
                language: sample.language || 'javascript',
                ...sample.options
            })
        );

        return Promise.all(promises);
    }

    /**
     * Get user profile information
     * 
     * @returns {Promise<Object>} User profile data
     */
    async getUserProfile() {
        try {
            const response = await this.client.get('/api/auth/profile');
            return response.data;
        } catch (error) {
            throw new CodeGuardianError(`Failed to get user profile: ${error.message}`);
        }
    }

    /**
     * Get API usage statistics
     * 
     * @returns {Promise<Object>} Usage statistics
     */
    async getUsageStats() {
        try {
            const response = await this.client.get('/api/reviews/stats');
            return response.data;
        } catch (error) {
            throw new CodeGuardianError(`Failed to get usage stats: ${error.message}`);
        }
    }

    /**
     * Test API connection
     * 
     * @returns {Promise<boolean>} True if connection is successful
     */
    async testConnection() {
        try {
            const response = await this.client.get('/health');
            return response.data.status === 'healthy';
        } catch (error) {
            return false;
        }
    }

    /**
     * Parse API response into AnalysisResult
     * @private
     */
    _parseAnalysisResult(response, analysisTime) {
        const comments = response.comments || [];
        const issues = comments.map(comment => new CodeIssue({
            lineNumber: comment.line_number || 0,
            column: comment.column || 0,
            severity: comment.severity || 'low',
            category: comment.category || 'general',
            message: comment.message || '',
            suggestion: comment.suggestion,
            fix: comment.fix
        }));

        return new AnalysisResult({
            overallScore: response.overall_score || 0,
            securityScore: response.security_score || 0,
            performanceScore: response.performance_score || 0,
            maintainabilityScore: response.maintainability_score || 0,
            issues,
            suggestions: response.suggestions || [],
            analysisTime
        });
    }

    /**
     * Detect programming language from file extension
     * @private
     */
    _detectLanguage(filePath) {
        const ext = path.extname(filePath).toLowerCase();
        const extensionMap = {
            '.js': 'javascript',
            '.ts': 'typescript',
            '.jsx': 'javascript',
            '.tsx': 'typescript',
            '.py': 'python',
            '.java': 'java',
            '.cpp': 'cpp',
            '.c': 'c',
            '.cs': 'csharp',
            '.go': 'go',
            '.rs': 'rust',
            '.php': 'php',
            '.rb': 'ruby'
        };
        return extensionMap[ext] || 'text';
    }
}

/**
 * Convenience functions for quick analysis
 */

/**
 * Quick analysis function
 * 
 * @param {string} code - Source code to analyze
 * @param {string} apiKey - API key
 * @param {Object} [options={}] - Analysis options
 * @returns {Promise<AnalysisResult>} Analysis results
 */
async function analyzeCode(code, apiKey, options = {}) {
    const client = new CodeGuardianClient({ apiKey });
    return client.analyzeCode(code, options);
}

/**
 * Quick file analysis function
 * 
 * @param {string} filePath - Path to file
 * @param {string} apiKey - API key
 * @param {Object} [options={}] - Analysis options
 * @returns {Promise<AnalysisResult>} Analysis results
 */
async function analyzeFile(filePath, apiKey, options = {}) {
    const client = new CodeGuardianClient({ apiKey });
    return client.analyzeFile(filePath, options);
}

/**
 * Quick code explanation function
 * 
 * @param {string} code - Source code to explain
 * @param {string} apiKey - API key
 * @param {Object} [options={}] - Explanation options
 * @returns {Promise<ExplanationResult>} Explanation results
 */
async function explainCode(code, apiKey, options = {}) {
    const client = new CodeGuardianClient({ apiKey });
    return client.explainCode(code, options);
}

// Export classes and functions
module.exports = {
    CodeGuardianClient,
    CodeIssue,
    AnalysisResult,
    ExplanationResult,
    Severity,
    CodeGuardianError,
    AuthenticationError,
    RateLimitError,
    AnalysisError,
    analyzeCode,
    analyzeFile,
    explainCode
};

// Example usage
if (require.main === module) {
    async function example() {
        const apiKey = 'your-api-key-here';
        const client = new CodeGuardianClient({ apiKey });

        // Test connection
        const connected = await client.testConnection();
        if (connected) {
            console.log('✅ Connected to CodeGuardian API');
        } else {
            console.log('❌ Connection failed');
            return;
        }

        // Analyze some code
        const code = `
function unsafeFunction(userInput) {
    eval(userInput); // Security vulnerability!
    return "done";
}
        `;

        try {
            const result = await client.analyzeCode(code, { language: 'javascript' });

            console.log(`Overall Score: ${result.overallScore}/100`);
            console.log(`Security Score: ${result.securityScore}/100`);
            console.log(`Issues Found: ${result.issues.length}`);

            result.issues.forEach(issue => {
                console.log(`  Line ${issue.lineNumber}: ${issue.message}`);
                if (issue.suggestion) {
                    console.log(`    Suggestion: ${issue.suggestion}`);
                }
            });

            // Get explanation
            const explanation = await client.explainCode(code);
            console.log(`\nExplanation: ${explanation.description}`);
            console.log(`Suggestions: ${explanation.suggestions}`);

        } catch (error) {
            console.error('Error:', error.message);
        }
    }

    example();
}

File: integrations/api-sdks/python/codeguardian_sdk.py
"""
CodeGuardian Python SDK
Official Python client library for CodeGuardian API
"""

import requests
import json
import time
from typing import Dict, List, Optional, Any, Union
from dataclasses import dataclass
from enum import Enum
import asyncio
import aiohttp
from pathlib import Path


class Severity(Enum):
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


@dataclass
class CodeIssue:
    line_number: int
    column: int
    severity: Severity
    category: str
    message: str
    suggestion: Optional[str] = None
    fix: Optional[Dict[str, str]] = None


@dataclass
class AnalysisResult:
    overall_score: int
    security_score: int
    performance_score: int
    maintainability_score: int
    issues: List[CodeIssue]
    suggestions: List[str]
    analysis_time: float


@dataclass
class ExplanationResult:
    description: str
    suggestions: str
    reasoning: str
    examples: List[str]


class CodeGuardianError(Exception):
    """Base exception for CodeGuardian SDK"""
    pass


class AuthenticationError(CodeGuardianError):
    """Raised when API authentication fails"""
    pass


class RateLimitError(CodeGuardianError):
    """Raised when API rate limit is exceeded"""
    pass


class AnalysisError(CodeGuardianError):
    """Raised when code analysis fails"""
    pass


class CodeGuardianClient:
    """
    CodeGuardian API Client
    
    Provides access to CodeGuardian's AI-powered code analysis and mentorship features.
    
    Example:
        client = CodeGuardianClient(api_key="your-api-key")
        result = client.analyze_code("print('hello world')", language="python")
        print(f"Score: {result.overall_score}/100")
    """
    
    def __init__(
        self,
        api_key: str,
        api_url: str = "https://codeguardian-api.onrender.com",
        timeout: int = 30,
        max_retries: int = 3
    ):
        """
        Initialize CodeGuardian client
        
        Args:
            api_key: Your CodeGuardian API key
            api_url: API base URL (default: production URL)
            timeout: Request timeout in seconds
            max_retries: Maximum number of retry attempts
        """
        self.api_key = api_key
        self.api_url = api_url.rstrip('/')
        self.timeout = timeout
        self.max_retries = max_retries
        
        self.session = requests.Session()
        self.session.headers.update({
            'Authorization': f'Bearer {api_key}',
            'Content-Type': 'application/json',
            'User-Agent': 'CodeGuardian-Python-SDK/1.0.0'
        })
    
    def analyze_code(
        self,
        code: str,
        language: str = "python",
        filename: Optional[str] = None,
        mentorship: bool = True,
        security_focus: bool = True,
        **options
    ) -> AnalysisResult:
        """
        Analyze code for issues, security vulnerabilities, and improvements
        
        Args:
            code: Source code to analyze
            language: Programming language (python, javascript, java, etc.)
            filename: Optional filename for context
            mentorship: Enable AI mentorship explanations
            security_focus: Focus on security vulnerabilities
            **options: Additional analysis options
            
        Returns:
            AnalysisResult with scores, issues, and suggestions
            
        Raises:
            AnalysisError: If analysis fails
            AuthenticationError: If API key is invalid
            RateLimitError: If rate limit is exceeded
        """
        payload = {
            'code': code,
            'language': language,
            'fileName': filename,
            'options': {
                'mentorship': mentorship,
                'securityFocus': security_focus,
                **options
            }
        }
        
        start_time = time.time()
        response = self._make_request('POST', '/api/reviews/analyze', json=payload)
        analysis_time = time.time() - start_time
        
        return self._parse_analysis_result(response, analysis_time)
    
    def analyze_file(
        self,
        file_path: Union[str, Path],
        language: Optional[str] = None,
        **options
    ) -> AnalysisResult:
        """
        Analyze a code file
        
        Args:
            file_path: Path to the code file
            language: Programming language (auto-detected if not provided)
            **options: Additional analysis options
            
        Returns:
            AnalysisResult with analysis results
        """
        file_path = Path(file_path)
        
        if not file_path.exists():
            raise FileNotFoundError(f"File not found: {file_path}")
        
        with open(file_path, 'r', encoding='utf-8') as f:
            code = f.read()
        
        if language is None:
            language = self._detect_language(file_path)
        
        return self.analyze_code(
            code=code,
            language=language,
            filename=str(file_path),
            **options
        )
    
    def analyze_directory(
        self,
        directory_path: Union[str, Path],
        recursive: bool = True,
        file_patterns: Optional[List[str]] = None,
        **options
    ) -> Dict[str, AnalysisResult]:
        """
        Analyze all code files in a directory
        
        Args:
            directory_path: Path to the directory
            recursive: Analyze subdirectories recursively
            file_patterns: File patterns to include (e.g., ['*.py', '*.js'])
            **options: Additional analysis options
            
        Returns:
            Dictionary mapping file paths to analysis results
        """
        directory_path = Path(directory_path)
        
        if not directory_path.exists():
            raise FileNotFoundError(f"Directory not found: {directory_path}")
        
        if file_patterns is None:
            file_patterns = [
                '*.py', '*.js', '*.ts', '*.java', '*.cpp', '*.c',
                '*.cs', '*.go', '*.rs', '*.php', '*.rb'
            ]
        
        files = []
        for pattern in file_patterns:
            if recursive:
                files.extend(directory_path.rglob(pattern))
            else:
                files.extend(directory_path.glob(pattern))
        
        results = {}
        for file_path in files:
            try:
                result = self.analyze_file(file_path, **options)
                results[str(file_path)] = result
            except Exception as e:
                print(f"Error analyzing {file_path}: {e}")
        
        return results
    
    def explain_code(
        self,
        code: str,
        language: str = "python",
        mentorship_level: str = "detailed"
    ) -> ExplanationResult:
        """
        Get AI explanation and mentorship for code
        
        Args:
            code: Source code to explain
            language: Programming language
            mentorship_level: Level of detail (basic, detailed, expert)
            
        Returns:
            ExplanationResult with description, suggestions, and reasoning
        """
        payload = {
            'code': code,
            'language': language,
            'mentorshipLevel': mentorship_level
        }
        
        response = self._make_request('POST', '/api/reviews/explain', json=payload)
        
        return ExplanationResult(
            description=response.get('description', ''),
            suggestions=response.get('suggestions', ''),
            reasoning=response.get('reasoning', ''),
            examples=response.get('examples', [])
        )
    
    def get_suggestions(
        self,
        code: str,
        language: str = "python",
        context: Optional[str] = None
    ) -> List[str]:
        """
        Get improvement suggestions for code
        
        Args:
            code: Source code
            language: Programming language
            context: Additional context for suggestions
            
        Returns:
            List of improvement suggestions
        """
        payload = {
            'code': code,
            'language': language,
            'context': context
        }
        
        response = self._make_request('POST', '/api/reviews/suggestions', json=payload)
        return response.get('suggestions', [])
    
    def check_security(
        self,
        code: str,
        language: str = "python"
    ) -> List[CodeIssue]:
        """
        Check code for security vulnerabilities
        
        Args:
            code: Source code to check
            language: Programming language
            
        Returns:
            List of security issues found
        """
        result = self.analyze_code(
            code=code,
            language=language,
            security_focus=True,
            mentorship=False
        )
        
        # Filter for security-related issues
        security_issues = [
            issue for issue in result.issues
            if 'security' in issue.category.lower()
        ]
        
        return security_issues
    
    def get_user_profile(self) -> Dict[str, Any]:
        """Get user profile information"""
        return self._make_request('GET', '/api/auth/profile')
    
    def get_usage_stats(self) -> Dict[str, Any]:
        """Get API usage statistics"""
        return self._make_request('GET', '/api/reviews/stats')
    
    def test_connection(self) -> bool:
        """Test API connection"""
        try:
            response = self._make_request('GET', '/health')
            return response.get('status') == 'healthy'
        except Exception:
            return False
    
    def _make_request(
        self,
        method: str,
        endpoint: str,
        **kwargs
    ) -> Dict[str, Any]:
        """Make HTTP request with retry logic"""
        url = f"{self.api_url}{endpoint}"
        
        for attempt in range(self.max_retries + 1):
            try:
                response = self.session.request(
                    method=method,
                    url=url,
                    timeout=self.timeout,
                    **kwargs
                )
                
                if response.status_code == 401:
                    raise AuthenticationError("Invalid API key")
                elif response.status_code == 429:
                    if attempt < self.max_retries:
                        time.sleep(2 ** attempt)  # Exponential backoff
                        continue
                    raise RateLimitError("Rate limit exceeded")
                elif response.status_code >= 400:
                    raise CodeGuardianError(f"API error: {response.status_code}")
                
                return response.json()
                
            except requests.exceptions.RequestException as e:
                if attempt < self.max_retries:
                    time.sleep(1)
                    continue
                raise CodeGuardianError(f"Request failed: {e}")
        
        raise CodeGuardianError("Max retries exceeded")
    
    def _parse_analysis_result(
        self,
        response: Dict[str, Any],
        analysis_time: float
    ) -> AnalysisResult:
        """Parse API response into AnalysisResult"""
        comments = response.get('comments', [])
        issues = []
        
        for comment in comments:
            issues.append(CodeIssue(
                line_number=comment.get('line_number', 0),
                column=comment.get('column', 0),
                severity=Severity(comment.get('severity', 'low')),
                category=comment.get('category', 'general'),
                message=comment.get('message', ''),
                suggestion=comment.get('suggestion'),
                fix=comment.get('fix')
            ))
        
        return AnalysisResult(
            overall_score=response.get('overall_score', 0),
            security_score=response.get('security_score', 0),
            performance_score=response.get('performance_score', 0),
            maintainability_score=response.get('maintainability_score', 0),
            issues=issues,
            suggestions=response.get('suggestions', []),
            analysis_time=analysis_time
        )
    
    def _detect_language(self, file_path: Path) -> str:
        """Detect programming language from file extension"""
        extension_map = {
            '.py': 'python',
            '.js': 'javascript',
            '.ts': 'typescript',
            '.jsx': 'javascript',
            '.tsx': 'typescript',
            '.java': 'java',
            '.cpp': 'cpp',
            '.c': 'c',
            '.cs': 'csharp',
            '.go': 'go',
            '.rs': 'rust',
            '.php': 'php',
            '.rb': 'ruby'
        }
        return extension_map.get(file_path.suffix.lower(), 'text')


class AsyncCodeGuardianClient:
    """
    Async version of CodeGuardian client for high-performance applications
    """
    
    def __init__(
        self,
        api_key: str,
        api_url: str = "https://codeguardian-api.onrender.com",
        timeout: int = 30,
        max_retries: int = 3
    ):
        self.api_key = api_key
        self.api_url = api_url.rstrip('/')
        self.timeout = timeout
        self.max_retries = max_retries
        
        self.headers = {
            'Authorization': f'Bearer {api_key}',
            'Content-Type': 'application/json',
            'User-Agent': 'CodeGuardian-Python-SDK/1.0.0'
        }
    
    async def analyze_code(
        self,
        code: str,
        language: str = "python",
        **options
    ) -> AnalysisResult:
        """Async version of analyze_code"""
        payload = {
            'code': code,
            'language': language,
            'options': options
        }
        
        start_time = time.time()
        
        async with aiohttp.ClientSession() as session:
            response = await self._make_request(
                session, 'POST', '/api/reviews/analyze', json=payload
            )
        
        analysis_time = time.time() - start_time
        return self._parse_analysis_result(response, analysis_time)
    
    async def analyze_multiple(
        self,
        code_samples: List[Dict[str, str]]
    ) -> List[AnalysisResult]:
        """Analyze multiple code samples concurrently"""
        async with aiohttp.ClientSession() as session:
            tasks = []
            for sample in code_samples:
                task = self.analyze_code(
                    code=sample['code'],
                    language=sample.get('language', 'python'),
                    **sample.get('options', {})
                )
                tasks.append(task)
            
            return await asyncio.gather(*tasks)
    
    async def _make_request(
        self,
        session: aiohttp.ClientSession,
        method: str,
        endpoint: str,
        **kwargs
    ) -> Dict[str, Any]:
        """Make async HTTP request"""
        url = f"{self.api_url}{endpoint}"
        
        async with session.request(
            method=method,
            url=url,
            headers=self.headers,
            timeout=aiohttp.ClientTimeout(total=self.timeout),
            **kwargs
        ) as response:
            if response.status == 401:
                raise AuthenticationError("Invalid API key")
            elif response.status == 429:
                raise RateLimitError("Rate limit exceeded")
            elif response.status >= 400:
                raise CodeGuardianError(f"API error: {response.status}")
            
            return await response.json()
    
    def _parse_analysis_result(
        self,
        response: Dict[str, Any],
        analysis_time: float
    ) -> AnalysisResult:
        """Parse API response into AnalysisResult"""
        # Same implementation as sync version
        comments = response.get('comments', [])
        issues = []
        
        for comment in comments:
            issues.append(CodeIssue(
                line_number=comment.get('line_number', 0),
                column=comment.get('column', 0),
                severity=Severity(comment.get('severity', 'low')),
                category=comment.get('category', 'general'),
                message=comment.get('message', ''),
                suggestion=comment.get('suggestion'),
                fix=comment.get('fix')
            ))
        
        return AnalysisResult(
            overall_score=response.get('overall_score', 0),
            security_score=response.get('security_score', 0),
            performance_score=response.get('performance_score', 0),
            maintainability_score=response.get('maintainability_score', 0),
            issues=issues,
            suggestions=response.get('suggestions', []),
            analysis_time=analysis_time
        )


# Convenience functions
def analyze_code(code: str, api_key: str, **kwargs) -> AnalysisResult:
    """Quick analysis function"""
    client = CodeGuardianClient(api_key)
    return client.analyze_code(code, **kwargs)


def analyze_file(file_path: str, api_key: str, **kwargs) -> AnalysisResult:
    """Quick file analysis function"""
    client = CodeGuardianClient(api_key)
    return client.analyze_file(file_path, **kwargs)


def explain_code(code: str, api_key: str, **kwargs) -> ExplanationResult:
    """Quick code explanation function"""
    client = CodeGuardianClient(api_key)
    return client.explain_code(code, **kwargs)


# Example usage
if __name__ == "__main__":
    # Example usage
    api_key = "your-api-key-here"
    client = CodeGuardianClient(api_key)
    
    # Test connection
    if client.test_connection():
        print("✅ Connected to CodeGuardian API")
    else:
        print("❌ Connection failed")
        exit(1)
    
    # Analyze some code
    code = """
def unsafe_function(user_input):
    exec(user_input)  # Security vulnerability!
    return "done"
    """
    
    result = client.analyze_code(code, language="python")
    
    print(f"Overall Score: {result.overall_score}/100")
    print(f"Security Score: {result.security_score}/100")
    print(f"Issues Found: {len(result.issues)}")
    
    for issue in result.issues:
        print(f"  Line {issue.line_number}: {issue.message}")
        if issue.suggestion:
            print(f"    Suggestion: {issue.suggestion}")
    
    # Get explanation
    explanation = client.explain_code(code)
    print(f"\nExplanation: {explanation.description}")
    print(f"Suggestions: {explanation.suggestions}")

